segmentstore/server/host/src/main/java/io/pravega/segmentstore/server/host/handler/PravegaRequestProcessor.java

```diff
@@ -704,11 +718,162 @@ public void readTable(final WireCommands.ReadTable readTable) {
+        tableStore.keyIterator(segment, state, TIMEOUT)
+                  .thenCompose(itr -> Futures.loop(
+                          canContinue::get,
+                          itr::getNext,
+                          e -> {
+                              if (e == null) {
+                                  canContinue.set(false);
+                              } else {
+                                  // store all keys.
+                                  keys.addAll(e.getEntries());
+                                  // update the continuation token.
+                                  ArrayView lastState = e.getState();
+                                  continuationToken.set(ByteBuffer.wrap(lastState.array(), lastState.arrayOffset(), lastState.getLength()));
+                                  // update the keycount and msgSize.
+                                  keyCount.addAndGet(e.getEntries().size());
+                                  msgSize.addAndGet(getTableKeyBytes(segment, e));
+
+                                  if (keyCount.get() >= suggestedKeyCount || msgSize.get() >= ITERATOR_RESPONSE_SIZE_THRESHOLD_BYTES) {
+                                      // set the continuation flag to false if we have read more than the suggested count or
+                                      // if we have read more the the response threshold bytes.
+                                      canContinue.set(false);
+                                  }
+                              }
+                          }, executor))
+                  .thenAccept(v -> {
+                      log.debug(getTableKeys.getRequestId(), "{} keys obtained for GetTableKeys request.", keys.size());
+                      List<WireCommands.TableKey> wireCommandKeys = keys.stream()
+                                                                        .map(k -> {
+                                                                            ArrayView keyArray = k.getKey();
+                                                                            return new WireCommands.TableKey(ByteBuffer.wrap(keyArray.array(),
+                                                                                                                             keyArray.arrayOffset(),
+                                                                                                                             keyArray.getLength()), k.getVersion());
+                                                                        })
+                                                                        .collect(toList());
+                      connection.send(new WireCommands.TableKeysIteratorItem(getTableKeys.getRequestId(), segment, wireCommandKeys, continuationToken.get()));
+                  }).exceptionally(e -> handleException(getTableKeys.getRequestId(), segment, operation, e));
+    }
+
+    @Override
+    public void getTableEntries(WireCommands.GetTableEntries getTableEntries) {
```

Comment:
Can we try to factor out common code from this method and `getTableKeys`? They look awfully similar, save for a couple of differences.

Reply:
I have refactored the code.